lightning/src/util/scid_utils.rs

```diff
@@ -60,6 +60,176 @@ pub fn scid_from_parts(block: u64, tx_index: u64, vout_index: u64) -> Result<u64
 	Ok((block << 40) | (tx_index << 16) | vout_index)
 }
 
+/// LDK has multiple reasons to generate fake short channel ids:
+/// 1) zero-conf channels that don't have a confirmed channel id yet
+/// 2) phantom node payments, to get an scid for the phantom node's phantom channel
+pub(crate) mod fake_scid {
+	use bitcoin::hash_types::BlockHash;
+	use bitcoin::hashes::hex::FromHex;
+	use chain::keysinterface::{Sign, KeysInterface};
+	use util::chacha20::ChaCha20;
+	use util::scid_utils;
+
+	use core::convert::TryInto;
+	use core::ops::Deref;
+
+	const TEST_SEGWIT_ACTIVATION_HEIGHT: u32 = 0;
+	const MAINNET_SEGWIT_ACTIVATION_HEIGHT: u32 = 481_824;
+	const MAX_TX_INDEX: u32 = 2_500;
+	const MAX_NAMESPACES: u8 = 8; // We allocate 3 bits for the namespace identifier.
+	const NAMESPACE_ID_BITMASK: u8 = 0b111;
+
+	/// Fake scids are divided into namespaces, with each namespace having its own identifier between
+	/// [0..7]. This allows us to identify what namespace a fake scid corresponds to upon HTLC
+	/// receipt, and handle the HTLC accordingly. The namespace identifier is encrypted when encoded
+	/// into the fake scid.
+	#[derive(Copy, Clone)]
+	pub(super) enum Namespace {
+		Phantom,
+		// Coming soon: a variant for the zero-conf scid namespace
+	}
+
+	impl Namespace {
+		/// We generate "realistic-looking" random scids here, meaning the scid's block height is
+		/// between segwit activation and the current best known height, and the tx index and output
+		/// index are also selected from a "reasonable" range. We add this logic because it makes it
+		/// non-obvious at a glance that the scid is fake, e.g. if it appears in invoice route hints.
+		pub(super) fn get_fake_scid<Signer: Sign, K: Deref>(&self, highest_seen_blockheight: u32, genesis_hash: &BlockHash, fake_scid_rand_bytes: &[u8; 32], keys_manager: &K) -> u64
+			where K::Target: KeysInterface<Signer = Signer>,
+		{
+			// Ensure we haven't created a namespace that doesn't fit into the 3 bits we've allocated for
+			// namespaces.
+			assert!((*self as u8) < MAX_NAMESPACES);
+			const BLOCKS_PER_MONTH: u32 = 144 /* blocks per day */ * 30 /* days per month */;
+			let rand_bytes = keys_manager.get_secure_random_bytes();
+
+			let segwit_activation_height = segwit_activation_height(genesis_hash);
+			let mut valid_block_range = if highest_seen_blockheight > segwit_activation_height {
```

Comment:
Little confused by the use of "range" in this variable name. Wondering if this instead could be:

```rust
let mut blocks_since_segwit_activation = highest_seen_blockheight.saturating_sub(segwit_activation_height);

// ...

let fake_scid_height = segwit_activation_height + rand_for_height % (blocks_since_segwit_activation + 1);